drivers/ethernet/eth_stm32_hal.c

```diff
@@ -1288,38 +1116,105 @@ static int eth_init_api_v2(const struct device *dev)
 	tx_config.CRCPadCtrl = ETH_CRC_PAD_INSERT;
 
-	HAL_ETH_SetMDIOClockRange(heth);
+	/* prepare tx buffer header */
+	for (uint16_t i = 0; i < ETH_TXBUFNB; ++i) {
+		dma_tx_buffer_header[i].tx_buff.buffer = dma_tx_buffer[i];
+	}
+
+	return 0;
+}
+#endif /* CONFIG_ETH_STM32_HAL_API_V2 */
+
+static void set_mac_config(const struct device *dev, struct phy_link_state *state)
+{
+	struct eth_stm32_hal_dev_data *dev_data = dev->data;
+	ETH_HandleTypeDef *heth = &dev_data->heth;
+	HAL_StatusTypeDef hal_ret = HAL_OK;
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+	ETH_MACConfigTypeDef mac_config = {0};
 
 	HAL_ETH_GetMACConfig(heth, &mac_config);
 
-#if defined(CONFIG_ETH_STM32_AUTO_NEGOTIATION_ENABLE)
-	/* Auto Nego enabled */
-	get_auto_nego_speed_duplex(heth, &mac_config);
+	mac_config.DuplexMode =
+		PHY_LINK_IS_FULL_DUPLEX(state->speed) ? ETH_FULLDUPLEX_MODE : ETH_HALFDUPLEX_MODE;
 
-#else /* Auto Nego disabled */
-	mac_config.DuplexMode = IS_ENABLED(CONFIG_ETH_STM32_MODE_HALFDUPLEX) ? ETH_HALFDUPLEX_MODE
-									     : ETH_FULLDUPLEX_MODE;
-	mac_config.Speed = IS_ENABLED(CONFIG_ETH_STM32_SPEED_10M) ? ETH_SPEED_10M : ETH_SPEED_100M;
-#endif
+	mac_config.Speed = PHY_LINK_IS_SPEED_100M(state->speed) ? ETH_SPEED_100M : ETH_SPEED_10M;
 
 	hal_ret = HAL_ETH_SetMACConfig(heth, &mac_config);
 	if (hal_ret != HAL_OK) {
 		LOG_ERR("HAL_ETH_SetMACConfig: failed: %d", hal_ret);
 	}
+#else  /* CONFIG_ETH_STM32_HAL_API_V2 */
+	heth->Init.DuplexMode =
+		PHY_LINK_IS_FULL_DUPLEX(state->speed) ? ETH_MODE_FULLDUPLEX : ETH_MODE_HALFDUPLEX;
 
-	/* prepare tx buffer header */
-	for (uint16_t i = 0; i < ETH_TXBUFNB; ++i) {
-		dma_tx_buffer_header[i].tx_buff.buffer = dma_tx_buffer[i];
+	heth->Init.Speed = PHY_LINK_IS_SPEED_100M(state->speed) ? ETH_SPEED_100M : ETH_SPEED_10M;
+
+	hal_ret = HAL_ETH_ConfigMAC(heth, NULL);
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("HAL_ETH_ConfigMAC: failed: %d", hal_ret);
 	}
+#endif /* CONFIG_ETH_STM32_HAL_API_V2 */
+}
 
+static int eth_stm32_hal_start(const struct device *dev)
+{
+	struct eth_stm32_hal_dev_data *dev_data = dev->data;
+	ETH_HandleTypeDef *heth = &dev_data->heth;
+	HAL_StatusTypeDef hal_ret = HAL_OK;
+
+	LOG_DBG("Starting ETH HAL driver");
+
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
 	hal_ret = HAL_ETH_Start_IT(heth);
+#else
+	hal_ret = HAL_ETH_Start(heth);
+#endif
 	if (hal_ret != HAL_OK) {
 		LOG_ERR("HAL_ETH_Start{_IT} failed");
 	}
 
 	return 0;
 }
-#endif /* CONFIG_ETH_STM32_HAL_API_V2 */
+
+static int eth_stm32_hal_stop(const struct device *dev)
+{
+	struct eth_stm32_hal_dev_data *dev_data = dev->data;
+	ETH_HandleTypeDef *heth = &dev_data->heth;
+	HAL_StatusTypeDef hal_ret = HAL_OK;
+
+	LOG_DBG("Stopping ETH HAL driver");
+
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+	hal_ret = HAL_ETH_Stop_IT(heth);
+#else
+	hal_ret = HAL_ETH_Stop(heth);
+#endif
+	if (hal_ret != HAL_OK) {
+		/* Stopping the ETH HAL only fails, when it's already stopped. */
+		LOG_DBG("HAL_ETH_Stop{_IT} failed, already not running");
```

Comment:
Let's avoid saying the function "failed" - rather, it hasn't done anything because it wasn't necessary.
```suggestion
		/* HAL_ETH_Stop{_IT} returns HAL_ERROR only if ETH is already stopped */
		LOG_DBG("HAL_ETH_Stop{_IT} returned error (Ethernet is already stopped)");
```

Reply:
applied